idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,850 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES), and is copyright (c) 2018-2021
+# by the software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+import pandas as pd
+import pyomo.environ as pyo
+from pyomo.opt.base.solvers import OptSolver
+import os
+from itertools import combinations
+from abc import ABC, abstractmethod, abstractproperty
+
+
+class AbstractBidder(ABC):
+
+    """
+    The abstract class for all the bidder and self-schedulers.
+    """
+
+    @abstractmethod
+    def update_model(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def compute_bids(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def write_results(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def formulate_bidding_problem(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def record_bids(self, *args, **kwargs):
+        pass
+
+    @abstractproperty
```

Comment:
`abstractproperty` was depreciated in Python 3.3. https://docs.python.org/3/library/abc.html#abc.abstractproperty